src/servers/Server.cc

```diff
@@ -150,6 +152,17 @@ Server::doClientRead(const CommIoCbParams &io)
     case Comm::ENDFILE: // close detected by 0-byte read
         debugs(33, 5, io.conn << " closed?");
 
+        // whether there is an outstanding transaction that has consumed all input bytes
+        if (!pipeline.empty() && inBuf.isEmpty()) {
+            const auto context = pipeline.back();
+            if (const auto http = context->http) {
+                if (const auto conn = http->getConn()) {
+                    if (conn->mayNeedToReadMoreBody()) // whether the body is truncated
+                        http->al->cache.requestReadTimer.update();
+                }
+            }
+        }
```

Comment:
Basically this code block accounts for a single case - a truncated PUT request with Content-Length. We don't need to deal with a PUT request without Content-Length, because Squid treats such requests as bodyless requests and all subsequent bytes as the beginning of another request.
Note that RFC7230 does not [prohibit](https://datatracker.ietf.org/doc/html/rfc7230#section-3.3.2) (with a MUST requirement) PUT requests without Content-Length:
```
A user agent SHOULD send a Content-Length in a request message when
no Transfer-Encoding is sent and the request method defines a meaning
for an enclosed payload body.  
```

Reply:
> RFC7230 does not [prohibit](https://datatracker.ietf.org/doc/html/rfc7230#section-3.3.2) (with a MUST requirement) PUT requests without Content-Length

Squid treats requests without Content-Length (and no chunked encoding) as a request with Content-Length=0. This case is covered in [RFC7230](https://datatracker.ietf.org/doc/html/rfc7230#section-3.3.3):

```
  6.  If this is a request message and none of the above are true, then
       the message body length is zero (no message body is present).
```
So it looks like that Squid treats such requests correctly.